### PR TITLE
Best Practices: Add note about object/Any

### PR DESCRIPTION
**Pull request by srittau**
Part of #851

### CODE REVIEW

**Review by hauntsaninja — COMMENTED** [review]
I'm not sure I'd recommend using `from __future__ import annotations` for new syntax without at least having a note mentioning that doing so breaks runtime uses of annotations (and e.g. type aliases won't work).

The note doesn't need to be lengthy or scary, e.g. it could just link to https://mypy.readthedocs.io/en/stable/runtime_troubles.html or something, but I do think we should acknowledge this.

**Comment by srittau** [comment]
> I'm not sure I'd recommend using from __future__ import annotations for new syntax without at least having a note mentioning that doing so breaks runtime uses of annotations (and e.g. type aliases won't work).

I'm not sure I follow. What runtime uses break that work without the future import.

File: docs/source/best_practices.rst [hunk 1]
```diff
@@ -44,12 +62,21 @@ No::
...
 
-Using `Any`
------------
+Using ``Any`` and ``object``
+----------------------------
 
 Generally, use ``Any`` when a type cannot be expressed appropriately
 with the current type system or using the correct type is unergonomic.
 
+If a function accepts virtually any object as an argument, for example
```
Comment (JelleZijlstra) [inline docs/source/best_practices.rst:71]:
```suggestion
If a function accepts every possible object as an argument, for example
```
Avoid using the word "any" to describe how to use "object", it's going to be confusing.

**Comment by JelleZijlstra** [comment]
I personally would not recommend using `from __future__ import annotations`:

- As Shantanu says, it can break some runtime use cases. See e.g. David Foster's message in https://discuss.python.org/t/finding-edge-cases-for-peps-484-563-and-649-type-annotations/14314 .
- It's still unclear whether the `__future__` import will ever become the default, so people using it may have to migrate later to another subtly different version.

**Comment by hauntsaninja** [comment]
It's not just the edge cases, since we're currently recommending `from __future__ import annotations` as a way to backport new syntax (not just for forward references). So `typing.get_type_hints` and friends will break.

Suggested wording (based on mypy docs):

> Some code may have use cases for evaluation of annotations at runtime, for instance, through use of [`typing.get_type_hints`](https://docs.python.org/3/library/typing.html#typing.get_type_hints) or `eval`. If your annotation would raise an error when evaluated (say by using new-style unions on Python 3.9 and older), you may need to be careful. See [here](https://mypy.readthedocs.io/en/stable/runtime_troubles.html ) for more discussion on sharp edges.

**Comment by srittau** [comment]
I have removed the future import section for now.

**Comment by srittau** [comment]
Bumpity bump.

**Pull request merged**